flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcher.java

```diff
@@ -51,7 +50,7 @@
  *            the Flink data streams.
  * @param <KPH> The type of topic/partition identifier used by Kafka in the specific version.
  */
-public abstract class AbstractFetcher<T, KPH> {
+public abstract class AbstractFetcher<T, KPH> implements KafkaOffsetCommitter {
```

Comment:
Maybe use composition instead of inheritance here? `AbstractFetcher` as "being a" `KafkaOffsetCommitter` do not make sense to me when I say it out laud, while `AbstractFetcher` "having a"  `KafkaOffsetCommitter` sounds better.

Reply:
I agree that composition suits better here, or maybe even neither of both. However, the reality is that currently the offset committing logic is implemented tightly as part of the `AbstractFetcher`, sharing the same Kafka client for both fetching records and committing offsets. Decoupling that would require further refactoring, which I think is a bit out of scope for the current issue at hand.

I have been thinking that we should simply have two separate service implementations for offset committing and record fetching. If that happens, then neither composition or inheritance is required; offset committing and record fetching simply lives as two separate services.

What do you think?

Reply:
The result could also very well be that we should use this opportunity to refactor the vague dependencies between fetcher / consumer thread / consumer base, and include in this PR. I would not be against that.

Reply:
Probably you are right, however I haven't thought this through, but if you had and you feel that's a good opportunity to make some larger refactor I'm fine with that :)

Anyway, as we discussed offline, such refactor could be made in separate PR and here you could just implement `AbstractFetcher#doCommitInternalOffsetsToKafka` for the testing purposes. 

I don't mind if you choose one over the other approach :)